pkg/icingadb/cleanup.go

```diff
@@ -0,0 +1,124 @@
+	"github.com/icinga/icingadb/pkg/utils"
+	"golang.org/x/sync/errgroup"
+	"time"
+)
+
+// CleanupResult is the result of executing a cleanup routine and
+// stores how many rows were deleted and how many statements were executed for that.
+type CleanupResult struct {
+	Count  uint64
+	Rounds uint64
+}
+
+// CleanupStmt defines information needed to compose cleanup statements.
+type CleanupStmt struct {
+	Table string
+	PK    string
+	Where string
+}
+
+// Build assembles the cleanup statement for the specified database driver with the given limit.
+func (stmt *CleanupStmt) Build(driverName string, limit uint) string {
+	switch driverName {
+	case driver.MYSQL:
+		return fmt.Sprintf(`DELETE FROM %s WHERE %s LIMIT %d`, stmt.Table, stmt.Where, limit)
+	case driver.PGSQL:
+		return fmt.Sprintf(`WITH rows AS (
+SELECT %[1]s FROM %[2]s WHERE %[3]s LIMIT %[4]d
+)
+DELETE FROM %[2]s WHERE %[1]s IN (SELECT %[1]s FROM rows)`, stmt.PK, stmt.Table, stmt.Where, limit)
+	default:
+		panic("invalid database type")
+	}
+}
+
+// CleanupOlderThan deletes all rows with the specified statement that are older than the given time.
+// Deletes a maximum of as many rows per round as defined in count.
+// If the query execution time exceeds the duration defined in the grace period,
+// the count is halved and the query is rerun.
+// The count is reset once the query has run within the grace period.
+func (db *DB) CleanupOlderThan(ctx context.Context, stmt CleanupStmt, count uint, gracePeriod time.Duration, olderThan time.Time) (CleanupResult, error) {
+	var counter com.Counter
+	defer db.log(ctx, stmt.Build(db.DriverName(), 0), &counter).Stop()
+	var rounds com.Counter
+
+	affectedRows := make(chan int64, 1)
+	defer close(affectedRows)
+
+	for limit, stop := count, false; !stop; {
+		deleteCtx, cancelDeleteCtx := context.WithCancel(ctx)
+		g, ctx := errgroup.WithContext(ctx)
+
+		g.Go(func() error {
+			stmt := stmt.Build(db.DriverName(), limit)
+			rs, err := db.NamedExecContext(deleteCtx, stmt, cleanupWhere{types.UnixMilli(olderThan)})
+			if err != nil {
+				if !utils.IsContextCanceled(deleteCtx.Err()) {
+					return internal.CantPerformQuery(err, stmt)
+				}
+				// Returns nil if only deleteCtx was canceled.
+				return ctx.Err()
+			}
+
+			n, err := rs.RowsAffected()
+			if err != nil {
+				return err
+			}
+
+			select {
+			case affectedRows <- n:
+				return nil
+			case <-ctx.Done():
+				return ctx.Err()
+			}
+		})
+
+		g.Go(func() error {
+			select {
+			case n := <-affectedRows:
+				rounds.Inc()
+				counter.Add(uint64(n))
+
+				if n < int64(limit) {
+					stop = true
+				} else {
+					limit = count
+				}
+			case <-time.After(gracePeriod):
+				db.logger.Warnf(
+					"Query %q could not be executed within %s. Limit is halved and the query is executed again",
+					stmt.Build(db.DriverName(), limit),
+					gracePeriod,
+				)
+				rounds.Inc()
+				counter.Add(uint64(<-affectedRows))
+
+				limit = limit / 2
+				cancelDeleteCtx()
```

Comment:
This looks the wrong way around. If I didn't miss something, cancelling the context doesn't have an effect because the other goroutine only uses it before sending to the channel.

Reply:
Removed the grace period (see above).